cloudinit/util.py

```diff
@@ -1814,6 +1813,33 @@ def time_rfc2822():
     return ts
 
 
+def boottime():
+    """Use sysctlbyname(3) via ctypes to find kern.boottime
```

Comment:
Can you provide the raw output of running this boottime() function on FreeBSD?  as well as the  sysctl -n kern.boottime?

Reply:
i'm using this extract:

```python
import pprint
pp = pprint.PrettyPrinter(indent=4).pprint

def boottime():
    """Use sysctlbyname(3) via ctypes to find kern.boottime
    kern.boottime is of type struct timeval. Here we create a
    private class to easier unpack it.
    @return boottime: float to be compatible with linux
    """
    import ctypes

    NULL_BYTES = b"\x00"

    class timeval(ctypes.Structure):
        _fields_ = [
            ("tv_sec", ctypes.c_int64),
            ("tv_usec", ctypes.c_int64)
        ]
        def __repr__(self):
            return "({}, {})".format(self.tv_sec, self.tv_usec)

    libc = ctypes.CDLL('/lib/libc.so.7')
    size = ctypes.c_size_t()
    size.value = ctypes.sizeof(timeval)
    buf = timeval()
    if libc.sysctlbyname(b"kern.boottime" + NULL_BYTES, ctypes.byref(buf),
                         ctypes.byref(size), None, 0) != -1:
        pp(buf)
        return buf.tv_sec + buf.tv_usec / 1000000.0
    raise RuntimeError("Unable to retrieve kern.boottime on this system")

pp(boottime())
```

to get this:

```
root@ubuntu-2gb-nbg1-1:~ # sysctl kern.boottime ; python3 boottime.py ; sysctl kern.boottime
kern.boottime: { sec = 1575455174, usec = 690313 } Wed Dec  4 10:26:14 2019
(1575455174, 690313)
1575455174.690313
kern.boottime: { sec = 1575455174, usec = 690313 } Wed Dec  4 10:26:14 2019
root@ubuntu-2gb-nbg1-1:~ # 
```

and here's the sysctl in binary form:

```
root@ubuntu-2gb-nbg1-1:~ # sysctl -b kern.boottime | od -D
0000000        1575455174               0          690313               0
0000020
root@ubuntu-2gb-nbg1-1:~ # sysctl -b kern.boottime
Ɖ]
root@ubuntu-2gb-nbg1-1:~ # sysctl -b kern.boottime | od -D
0000000        1575455174               0          690313               0
0000020
root@ubuntu-2gb-nbg1-1:~ # 
```

Reply:
Thanks.  I think for now, instead of the integration test with the skipUnlessFreebsd; let's just mock boottime() call and test the non-linux path.

```
@mock.patch('cloudinit.util.boottime')
@mock.patch('cloudinit.util.os.path.exists')
@mock.patch('cloudinit.util.time.time')
def test_uptime_non_linux_path(self, m_time, m_exists, m_boottime):
    boottime = 1000.0
    uptime = 10.0
    m_boottime.return_value = boottime
    m_time.return_value = boottime + uptime
    m_exists.return_value = False
    result = util.uptime()
    self.assertEqual(str(uptime), result)
```

Reply:
I'd like to implement a unittest that mocks out the ctypes calls but I don't have time to implement that as an example right now and we can add a unittest in a separate PR once that's figured out.

Reply:
>  I think for now, instead of the integration test with the skipUnlessFreebsd; let's just mock boottime() call and test the non-linux path.

why not both?
we could use something like https://cirrus-ci.org/guide/FreeBSD/ this in addition to travis if we're serious about supporting FreeBSD for real…

Reply:
By "integration tests", I mean cloudinit/tests/cloud_tests/  which is far more involved than running a some tests against  a host;